### ISSUE TITLE
Add Controller Support for Hotkeys

### ISSUE THREAD

**Issue by PleaseForTheLoveOfGod** [Lime3DS/Lime-3DS-Emulator#119]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What feature are you suggesting?

Right now you can only bind hotkeys to keyboard keys, which is extremely annoying for people who play on a controller and away form the keyboard. It would be really helpful to bind hotkeys like swap screens, save state, load state and reset emulation to controller buttons.

### Why would this feature be useful?

It would make gameplay more fluid, and just overall improve the experience of using the emulator for people who play on controller, especially those that also play away from keyboard.

**Comment by TomTurbine** [comment]
+1

The number 1 feature request I have for this is configurable controller hotkeys. The ability to configure it to swap screen layouts or exit a game and select a new one on the controller or just exit out entirely all on the controller would be such a huge QoL improvement.

The number 2 feature request would be an auto-updater feature.

**Comment by zhen-huan-hu** [comment]
This would be really a life saver for Steam Deck.

**Comment by rrrcabral** [comment]
I want this too. especially to be able to switch screens by pressing Home/Guide on the Controller, or to configure another button to toggle Fullscreen,
Currently, to do this the user has to run an external secondary program, such as AntiMicroX

**Comment by mcthomas** [comment]
> This would be really a life saver for Steam Deck.

@zhen-huan-hu You can set a regular keyboard input binding in Azahar, and then open the Steam menu over ES-DE, navigate to the controller settings, and modify the layout for your controller to add an additional binding to a controller input that matches the keyboard input you bound in Azahar.

**Comment by weblue** [comment]
Would be great for my RP5; it has basic gamepad controls, but no extra buttons (other than home + back for android). Would also make the experience consistent with some of my other emulators.

**Comment by KAR0TTENP0WER** [comment]
I also would love that feature.
Right now i have to create a Macro within ds4windows to assign a controller button to "ctrl+z" so i can enable/disable emulation speed.
I prefer not using ds4windows since i encounter problems here and then with it running.

**Comment by carter-lilley** [comment]
+1, was following this for years previous to azahar. Its time.

**Comment by shinra358** [comment]
Oops, didn't see this at first because it's all the way at the end.
https://github.com/azahar-emu/azahar/issues/1160

**Comment by salemxvii** [comment]
+1 - please

**Comment by jesperbekkering** [comment]
+1!

**Comment by retrogamedesigner** [comment]
As someone who likes to have my mini PCs and mobile devices completely controllable with a gamepad for a console-like experience, I would love this feature (especially on desktop platforms).

**Comment by InJJWeTrust** [comment]
Hopefully the priority of this can be increased.

I am curious why this is still not a thing, as many people I know as well as myself have been waiting for this for years, with many considering it a basic/essential part of 3DS emulation. Needing to find another software or a janky 3-step keyboard shortcut to controller shortcut type of cobbled-together solution feels very last-resort.

Looking forward to hearing updates on the progress of this.

**Comment by maart999** [comment]
Is there any update to this? This seems pretty important

**Comment by DavidRGriswold** [comment]
Please remember, when demanding work on a feature, that this is a volunteer project with just a few active developers. By all means continue to express support for the feature, but try to avoid sounding entitled to getting whatever feature *you* want on the free emulator nobody is paid for working on. 

In this *particular* instance, at least two devs (including me) have made attempts at this feature over the last six months and run into major challenges or snags at both technical and UI design levels. You are not being ignored, this is simply harder than it sounds. Given that there *are* ways that users with controller-focused pc setups can solve this problem - steam input, joy2key, ds4windows, etc - it is still something we plan to add but is honestly *not* as high a priority as other changes that will enable things currently impossible, or fix bugs, or improve performance.

**Comment by carter-lilley** [comment]
I completely appreciate how difficult this is and have nothing but gratitude for the ongoing work of the many contributors that make not only Azahar but the broader emulation scene as a whole possible. Thank you for your time and effort. 

A lot of the people in this thread are moreso looking for updates - is there a PR or something to track these failed implementations you speak of? It would help put into perspective the work being done and the challenges surrounding it. (And may encourage others to contribute or hop in to help.)

You have a dozen users across two years in this thread that have been met with silence. I understand the team's priorities lie with breaking changes, performance, etc, but there is clearly interest in a solution that does not involve third-party software, which as far as I am concerned is not a solution.

**Comment by maart999** [comment]
My apologies, I did not mean to sound entitled and you are absolutely right. I really did just mean it as asking for an update as I am very grateful for all the work you and the team do for free, that is very admirable

**Comment by DavidRGriswold** [comment]
Thanks for the responses, sorry for being snippy. Let's be productive with this moment:

If a dev were to start working on gamepad-based hotkey support, there are a few questions that need to be answered that have caused us some false starts and might benefit from user thoughts. Please feel free to give your thoughts on these questions so we can think about the best path forward.

## Question 1 - How to represent hotkeys?

Should the hotkey implementation be based on the *connected controller* or based on the *mapped 3DS buttons*? Meaning, if we want to map, say, Swap Screens to a controller, should it be mapped to something like "Home + X" (3DS buttons) or "Xbox Controller 1: L3" (a particular connected controller)

*My Thoughts*: Eden (switch emulator) uses the first option. It would be easier to do, and would allow for switching controllers without needing to remap hotkeys (only 3DS inputs). But it means that if you have an extra non-mapped controller button (e.g. a back button) it would not be useable as a hotkey. So my instinct would be to lean to the second option.

## Question 2 - Connect to controller profiles?

Azahar has an existing controller profile feature, but hotkeys are currently treated as a completely separate thing. Assuming we choose the second option above, should hotkeys be saved within a specific controller profile? That feels logical, but would require significant restructuring.

## Question 3 - Allow arbitrary button chords, a single "Enable Hotkey" button, or neither?

Mapping controller buttons to hotkeys is straightforward when there are enough buttons available, but most controllers do not have significantly more buttons than the 3DS. Thus it seems that some form of button-chording feels necessary for this to make sense. Some emulators (Dolphin) have extremely flexible button-chording options, but implementing those would be a huge lift. A reasonable middle ground would be an "Enable Hotkey" button like that used by RetroArch, so users could map the Enable Hotkey to either an unused button (L3, R3, back button) or a lesser-used key (Select) and then create chords of the form Enable+OtherButton as hotkeys. Which of these options should we pursue?

## Question 4 - Every hotkey, or just certain common in-game ones?

Some of the hotkeys currently enabled in the desktop version of azahar, such as Swap Screens, Toggle Layout, End Emulation, Make Save State, etc, could be implemented at the in-game level relatively easily, and would not need to interact with the user interface to function. Others, however, such as Load File and Toggle Status Bar, would require interacting with the user interface, which complicates the code. Would it be acceptable to make it so only *certain* actions could be done with controller hotkeys?

**Comment by DavidRGriswold** [comment]
Here is a mockup I made of *one potential* interface we could use to add controller hotkeys. This particular mockup would be in the current "Input" tab on desktop, so it would be associated with existing input profiles, which is good, but is obviously limited in terms of which hotkeys it allows. It matches what is available on android, however. Please react with either a thumbs up or thumbs down for whether this would satisfy your needs, and feel free to elaborate.

<img width="764" height="480" alt="Image" src="https://github.com/user-attachments/assets/695086b1-40b5-4d9c-a137-2e72205193f8" />

**Comment by bmsuseluda** [comment]
for me this would be really good. I think the Hotkey enable should be optional. So if you have enough buttons you do not need to press button combinations.

**Comment by InJJWeTrust** [comment]
I agree, from my standpoint at least this would at the very least be an 80/20 solution that would solve the majority of people's immediate needs. It could then be further brushed up at a later date with a lower priority as deemed neccesary.

Love the fact that it is also largely in alignment with the current options on Android as well. 

Nice and clean! Great work!

**Comment by maart999** [comment]
Yeah I agree too, this seems like best of both worlds, great stuff!

**Comment by DavidRGriswold** [comment]
Proof of concept. Still very early days. 

https://github.com/user-attachments/assets/45bd91ed-314e-4f9a-a9f9-2c70436c00c9

**Comment by DavidRGriswold** [comment]
<img width="1776" height="1402" alt="Image" src="https://github.com/user-attachments/assets/f82958e6-0b6f-47ec-8e09-34434ea1dff2" /> Coming Along! Decided to go with generic two-button chords (or single-button of course) rather than an Enable button.

**Comment by carter-lilley** [comment]
You are doing incredible work! Contributors like you make it all possible! Exciting - tysm!

**Comment by InJJWeTrust** [comment]
@DavidRGriswold 
You're doing God's work my friend. 
Imho, the impact which this change will have for both existing as well as new potential users is quite large. Keep up the great work! 🙏

**Comment by DavidRGriswold** [comment]
I have a first draft of this feature made. It still does not support input profiles, so if you switch between different controllers it will require remapping for the new controller each time. It may also be unstable! But if you are willing to test and provide some feedback, you can find the PR here: https://github.com/azahar-emu/azahar/pull/1717 . If you are logged into Github, you should be able to see the builds for the PR (still building as of this writing, should be ready about an hour after this post) and download the one for your desktop platform of choice. I have only test on Mac as of this writing, but it *should* work with linux and windows as well.

**Comment by DavidRGriswold** [comment]
If anybody uses windows or linux and can test the current version of the PR, I would greatly appreciate it, *especially* if your controller registers d-pad presses as Hats rather than as Buttons when you map them. Mac OS X apparently *always* creates button presses for dpads (tested with four different controllers...), and my linux machine is being fussy, so I'm having a hard time testing that functionality. 

To test, download the PR, open preferences, and try to register a dpad button as a hotkey and see if it (a) lets you register it and (b) if the hotkey actually fires, and let me know here or in the PR itself.

**Comment by retrogamedesigner** [comment]
Had issues with "Hat" input in Controller Hotkeys on build from daa37df on Linux Debian Trixie. It could only be assigned if another button was also simultaneously held while setting the hotkey. Also, any hat input was displayed as "Hat0 Centered," regardless of which direction on the d-pad was pressed. I'm currently building a3d518b and I will test it as soon as possible.

**Comment by retrogamedesigner** [comment]
I'm happy to report that everything seems to be working correctly on a3d518b build. The "Hat" can be assigned as a standalone hotkey with each direction being it's own unique input. The Hat directions can also be combined with other inputs as "combo" hotkeys. For example, Toggle Screen Layout: Button 8 + Hat 0 left

**Comment by DavidRGriswold** [comment]
Yay! Good news. I think analog axes should be working now too , both sticks and triggers.  I don't have a lot of time to actually play anything over the next few days, but if anybody is going to have a decently long session and could test the controller hotkeys while they play that'd be great! Be on the lookout for...

- consistent functionality of the controller hotkeys
- other inputs working as expected
- crashes or freezes

**Comment by Johann-Dark** [comment]
Hi, wanted to report that in my case with the extra features 0.7 build, the game (MH3U) crashes when I load the save - I'm using steam to lauch the emulator but I don't have any feature enabled
